### PR TITLE
Enable the crashreporting screen for sentry

### PR DESCRIPTION
**Pull request by strseb**
## Description
Before we send crash -reports we need user consent. This PR utilizes our old crash screen and retro-fits it into the new sentry flow. 
- Moving the UI from being it's own command into a "Screen" 
- Adding a Flag in the Sentry Reporter whether consent has already been given for this execution
- Adding a Filter in task Sentry: 
- Checks on every outgoing request if it contains a crashreport, if so it needs consent
- If not (just technical data), it will take the telemetry flag into account.  
- Only if the filter is passed, stuff will be send. 

## Next Steps
-> After this PR landed we can rip out /crashreporter :)  https://github.com/mozilla-mobile/mozilla-vpn-client/pull/5143

## Reference

https://mozilla-hub.atlassian.net/browse/VPN-2823

## Checklist
    
- [x] Cool video
- [x] Tests

Screen Recording 2022-12-06 at 20

### CODE REVIEW

File: src/frontend/navigator.cpp [hunk 1]
```diff
@@ -126,6 +126,13 @@ ScreenData s_screens[] = {
         QVector<MozillaVPN::State>{MozillaVPN::StateMain},
         [](Navigator::Screen*) -> int8_t { return 0; },
         []() -> bool { return false; }),
+    ScreenData(
+        Navigator::Screen::ScreenCrashReporting,
+        Navigator::LoadPolicy::LoadTemporarily,
+        "qrc:/ui/screens/ScreenCrashReporting.qml",
+        QVector<MozillaVPN::State>{},
+        [](Navigator::Screen*) -> int8_t { return 0; },
```
Comment (bakulf) [inline src/frontend/navigator.cpp:134]: Replace this with:
```
        [](Navigator::Screen* requestedScreen) -> int8_t {
          return (requestedScreen &&
                  *requestedScreen == Navigator::Screen::ScreenCrashReporting)
                     ? 99
                     : -1;
        },
```

File: src/sentry/sentryadapter.cpp [hunk 2]
```diff
@@ -17,6 +17,7 @@
 #include "tasks/sentry/tasksentry.h"
 #include "taskscheduler.h"
 #include "mozillavpn.h"
+#include "frontend/navigator.h"
```
Comment (bakulf) [inline src/sentry/sentryadapter.cpp:20]: alphabetical order

File: src/sentry/sentryadapter.cpp [hunk 3]
```diff
@@ -150,3 +151,21 @@ void SentryAdapter::transportEnvelope(sentry_envelope_t* envelope,
   auto t = new TaskSentry(qt_owned_buffer);
   TaskScheduler::scheduleTask(t);
 }
+
+SentryAdapter::UserConsentResult SentryAdapter::hasCrashUploadConsent() {
```
Comment (bakulf) [inline src/sentry/sentryadapter.cpp:155]: if (!m_initialized) { return  ... ? }
Maybe: Q_ASSERT(m_initialized);

File: src/sentry/sentryadapter.h [hunk 4]
```diff
@@ -90,8 +90,43 @@ class SentryAdapter final : public QObject {
...
+
+  /**
+   * @brief Checks if the user gave consent to upload crashes
+   *
+   * @return UserConsentResult - Allowed/Forbidden in case we have gotten
+   * consent Can return "Pending", in which case a prompt has been fired to the
+   * UI to ask the user.
+   * - listen for userConsentChanged to be notified when that is done
+   */
+  UserConsentResult hasCrashUploadConsent();
```
Comment (bakulf) [inline src/sentry/sentryadapter.h:108]: `UserConsentRequest hasCrashUploadConsent() const` ?

File: src/sentry/sentryadapter.h [hunk 5]
```diff
@@ -90,8 +90,43 @@ class SentryAdapter final : public QObject {
...
+
+  /**
+   * @brief Signal that is fired whenever the consent changed, note: can sill be
+   * Pending.
+   *
+   * @return bool
+   */
+  Q_SIGNAL void userConsentChanged();
```
Comment (bakulf) [inline src/sentry/sentryadapter.h:116]:
```
signals:
  void userConsentChanged();
```

File: src/tasks/sentry/tasksentry.cpp [hunk 6]
```diff
@@ -21,6 +28,42 @@ TaskSentry::TaskSentry(const QByteArray& envelope) : Task("TaskSentry") {
...
+    return;
+  }
+
+  // This is a crash report - Make sure we have consent, or ask the user for it
+  auto consent = SentryAdapter::instance()->hasCrashUploadConsent();
+
+  if (consent == SentryAdapter::UserConsentResult::Pending) {
+    // The consent screen was now triggered;
+    // exit here and connect to the change event to re-run.
+    connect(SentryAdapter::instance(), &SentryAdapter::userConsentChanged, this,
```
Comment (bakulf) [inline src/tasks/sentry/tasksentry.cpp:51]: can be the dialog close? Can you get blocked here waiting for the user consent?

File: src/tasks/sentry/tasksentry.cpp [hunk 7]
```diff
@@ -32,7 +75,68 @@ void TaskSentry::run() {
...
+            // connect customer support logs with sentry :)
+            logger.debug() << "Sentry sent event (" << m_eventID << ")";
             emit completed();
           });
-}
\ No newline at end of file
+}
+
+bool TaskSentry::isCrashReport() {
+  if (m_checkedContent) {
```
Comment (bakulf) [inline src/tasks/sentry/tasksentry.cpp:86]: Can you have just 1 enum to cover all these variables?
```
enum State {
  Unknown,
  CrashReport,
  Ping,
};
```

File: src/tasks/sentry/tasksentry.cpp [hunk 8]
```diff
@@ -32,7 +75,68 @@ void TaskSentry::run() {
...
+  m_checkedContent = true;
+  return m_isCrashReport;
+}
+
+void TaskSentry::parseEnvelope() {
+  auto envelope = QString::fromUtf8(m_envelope);
+
+  // The Sentry Envelope is a basic format
+  // It is a series of JSON objects splited by lines.
+  QStringList objects = envelope.split("\n");
```
Comment (bakulf) [inline src/tasks/sentry/tasksentry.cpp:100]: we need a unit-test for this.

File: src/tasks/sentry/tasksentry.cpp [hunk 9]
```diff
@@ -32,7 +75,68 @@ void TaskSentry::run() {
...
+  if (objects.empty()) {
+    Q_UNREACHABLE();  // This really should not happen.
```
Comment (bakulf) [inline src/tasks/sentry/tasksentry.cpp:102]: you never know... handle this case too.

File: src/tasks/sentry/tasksentry.cpp [hunk 10]
```diff
@@ -32,7 +75,68 @@ void TaskSentry::run() {
...
+  }
+  // The fist line is always the header:
+  // a json object, containing the DSN and an event ID
+  auto header = objects.takeFirst();
+  if (!header.contains("\"event_id\":")) {
```
Comment (bakulf) [inline src/tasks/sentry/tasksentry.cpp:107]: Please, parse it as JSON:
- if it's not an object
- or the object does not contain event_id...

File: src/tasks/sentry/tasksentry.cpp [hunk 11]
```diff
@@ -32,7 +75,68 @@ void TaskSentry::run() {
...
+    // if there is no event-id, there will be no event following that
+    // therefore this can't be a crashreport.
+    // So we can stop parsing this blob. c:
+    return;
+  } else {
```
Comment (bakulf) [inline src/tasks/sentry/tasksentry.cpp:112]: No `else` after a `return`.

File: src/tasks/sentry/tasksentry.cpp [hunk 12]
```diff
@@ -32,7 +75,68 @@ void TaskSentry::run() {
...
+    m_eventID = doc["event_id"].toString();
+  }
+  // Each Line here is it's own json object.
+  // Usually in the form of
+  // {{ headerObject type,len }} \n
+  // {{ bodyObject meta}}
+  foreach (auto content, objects) {
+    auto doc = QJsonDocument::fromJson(content.toUtf8());
+    if (!doc.isObject()) {
+      logger.error() << "Invalid content read from the JSON file";
```
Comment (bakulf) [inline src/tasks/sentry/tasksentry.cpp:124]: should we stop parsing the rest?

**Review by bakulf — CHANGES_REQUESTED** [review]

File: src/tasks/sentry/tasksentry.cpp (hunk 6 above)
Comment (strseb) [reply src/tasks/sentry/tasksentry.cpp:51]: Can only be closed by choosing an option.

File: src/tasks/sentry/tasksentry.cpp (hunk 8 above)
Comment (strseb) [reply src/tasks/sentry/tasksentry.cpp:101]: Added them :)

File: src/ui/screens/ScreenCrashReporting.qml [hunk 13]
```diff
@@ -0,0 +1,106 @@
...
+import QtQuick.Controls 2.14
+import QtQuick.Layouts 1.14
+
+import Mozilla.VPN 1.0
+import components 0.1
+
+VPNFlickable {
+    id: vpnFlickable
+
+    readonly property bool isMobile: window.fullScreenRequired()
```
Comment (MattLichtenstein) [inline src/ui/screens/ScreenCrashReporting.qml:15]: typo: should be `window.fullscreenRequired()` (lowercase s)

File: src/sentry/sentryadapter.cpp [hunk 14]
```diff
@@ -150,3 +151,22 @@ void SentryAdapter::transportEnvelope(sentry_envelope_t* envelope,
   auto t = new TaskSentry(qt_owned_buffer);
   TaskScheduler::scheduleTask(t);
 }
+
+SentryAdapter::UserConsentResult SentryAdapter::hasCrashUploadConsent() const {
+  Q_ASSERT(m_initialized);
+  // We have not yet asked the user - let's do that.
+  if (m_userConsent == UserConsentResult::Pending) {
+    Navigator::instance()->requestScreen(Navigator::ScreenCrashReporting);
```
Comment (bakulf) [inline src/sentry/sentryadapter.cpp:159]: emit a signal. And use the signal in the Navigator class.

File: tests/unit/testtasksentry.cpp [hunk 15]
```diff
@@ -0,0 +1,169 @@
...
+
+#include "testtasksentry.h"
+#include "../../src/settingsholder.h"
+
+#include <QTextStream>
+
+#include "../../src/tasks/sentry/tasksentry.h"
+#include "helper.h"
+
+void TestTaskSentry::envelopeWithJustAnAttachment() {
```
Comment (bakulf) [inline tests/unit/testtasksentry.cpp:13]: use the `_data()` slot for this. See other tests.

**Review by bakulf — CHANGES_REQUESTED** [review]
Can you add a functional test too?

File: src/sentry/sentryadapter.cpp (hunk 14 above)
Comment (strseb) [reply src/sentry/sentryadapter.cpp]: done :)

File: tests/unit/testtasksentry.cpp (hunk 15 above)
Comment (strseb) [reply tests/unit/testtasksentry.cpp]: done :)

**Comment by strseb** [comment]
> Can you add a functional test too? 
No at this stage i don't think we can write a meaningful functional test. 
Until we have removed the Sentry-Keys from the client (and do that via guardian) - the whole code is dormant and would require mocking the whole backend flow.

**Comment by birdsarah** [comment]
How do I test this? When I tried running this locally (on macos) I went to the developer menu and clicked Test Crash Reporter. After the countdown the app crashed but nothing happened:

image

I tried both logged in and not.

Is it because I don't have sentry keys?

**Comment by birdsarah** [comment]
(I tried looking in the 1password shared directory and the sentry keys weren't there.)

**Comment by strseb** [comment]
You can find the Sentry-Keys in the VPN-Sentry project :) 
If they are not set the whole compilation of that part is skipped.

**Pull request merged**